keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -0,0 +1,593 @@
+unconditionally would break some clusters, so this is also not an
+option.
+
+To make this work, `rest.InClusterConfig()` would have to introspect
+its own local network setup (eg, using `net.Interfaces()`) to see if
+it would be able to reach `KUBERNETES_SERVICE_HOST`, and then use an
+alternative (`KUBERNETES_SERVICE_HOST_SECONDARY` or
+`kubernetes.default.svc.cluster.local`) only if the primary apiserver
+IP is unreachable from that pod.
+
+There was [discussion in an earlier KEP] about whether it is actually
+safe to add new `KUBERNETES_*` environment variables, given that we
+don't restrict users from using variables with such names for their
+own purposes...)
+
+It's possible that there is no solution that is nice enough that it's
+actually worth trying to make it work for the admittedly-edge-case-y
+scenario of wrong-single-stack pods...
+
+[discussion in an earlier KEP]: https://github.com/kubernetes/enhancements/pull/1418#discussion_r405636588
+
+### Service and EndpointSlice Reconciling
+
+If the dual-stack apiserver changes were to take effect in a cluster
+all at once when the feature went to beta, then this could cause
+Service/EndpointSlice flapping in dual-stack clusters during the
+upgrade when the `DualStackAPIServer` feature gate becomes enabled,
+because the older apiservers would be trying to enforce a single-stack
+`kubernetes.default` service, while the newer apiservers would be
+trying to make it be dual-stack.
+
+(Note that the problem is only with `Service` and `EndpointSlice`, not
+with `Endpoints` because `Endpoints` is inherently single-stack, so
+both old and new apiservers will always agree on which IPs should be
+listed there.)
+
+To prevent this, we can make some of the dual-stack apiserver support
+code be active even when the feature gate is disabled; the feature
+gate will control whether an apiserver in a dual-stack cluster tries
+to advertise *itself* as dual-stack, but not whether it allows *other*
+apiservers to advertise themselves as dual-stack. Thus:
+
+  - API server does not know about `DualStackAPIServer` (eg,
+    Kubernetes 1.21, regardless of single/dual stack configuration):
+
+      - Advertises its own single-stack endpoint IP to etcd
+
+      - Forces the `kubernetes` Service to be single-stack when it
+        starts up.
+
+      - Syncs Endpoints and the primary-IP-family EndpointSlice.
+        Mistakenly ignores the secondary-IP-family EndpointSlice
+        ([kubernetes #101070]) and so could leave behind a stale
+        secondary-IP-family EndpointSlice on downgrade from dual-stack
+        apiserver.
+
+  - API server knows about `DualStackAPIServer` but either the feature
+    is not enabled on this apiserver, or else this apiserver has a
+    single-stack configuration:
+
+      - Advertises its own *single-stack* endpoint IP to etcd
+
+      - Forces the `kubernetes` Service to single-stack if there are
+        no dual-stack apiservers in etcd. Leaves the Service unchanged
```

Comment:
Based on what's in etcd. If you're using [`--endpoint-reconciler-type=master-count`](https://github.com/kubernetes/kubernetes/blob/master/pkg/controlplane/reconcilers/instancecount.go), it would see that some other apiserver had created dual-stack EndpointSlices and react to that. (Although this would be flaky... maybe we want to disallow `--endpoint-reconciler-type=master-count` with dual-stack apiservers.) If you're using [`--endpoint-reconciler-type=lease`](https://github.com/kubernetes/kubernetes/blob/master/pkg/controlplane/reconcilers/lease.go) then the endpoint IPs are stored in etcd separately from the Endpoints/EndpointSlice objects... although we will need to figure out how to change that code in a way that doesn't break existing single-stack apiservers...